probdists/Triangulardistribution.py

```diff
@@ -0,0 +1,199 @@
+    like a triangle
+    Note: a <= mode <= b
+
+    Attributes:
+
+        a (float): the minimum lower limit value
+        b (float): the maximum upper limit value
+        mode (float): the mode, where min <= mode <= max
+
+        mean (float): the mean value of the distribution
+        stdev (float): the standard deviation of the distribution
+
+    """
+
+    def __init__(self, a=0, b=1, mode=0.5):
+        if mode < a or b < mode:
+            raise ValueError
+
+        self.a = a
+        self.b = b
+        self.mode = mode
+
+        Distribution.__init__(self, self.calculate_mean(),
+                              self.calculate_stdev())
+
+    def calculate_mean(self, round_to=2):
+        """
+        Method to calculate the mean from the min, max and mode
+
+        Args:
+            round_to (int): Round the mean value. Defaults to 2.
+
+        Returns:
+            float: mean of the data set
+        """
+
+        self.mean = 1/3 * (self.a + self.b + self.mode)
+
+        return round(self.mean, round_to)
+
+    def calculate_stdev(self, round_to=2):
+        """
+        Method to calculate the standard deviation from the min, max and mode
+
+        Args:
+            round_to (int): Round the mean value. Defaults to 2.
+
+        Returns:
+            float: standard deviation of the data set
+        """
+
+        summation = (self.a ** 2) + (self.b ** 2) + (self.mode ** 2) - \
+                    (self.a * self.b) - (self.a * self.mode) - \
+                    (self.b * self.mode)
+        variance = summation / 18
+        self.stdev = math.sqrt(variance)
+
+        return round(self.stdev, round_to)
+
+    def replace_stats_with_data(self):
+        """ Method to calculate a, b, mode from the data set
+
+        Args:
+            None
+
+        Returns:
+            float: a, the minimum value
+            float: b, the maximum value
+            float: mode, the mode of the dataset
+        """
+        if not self.data:
+            # Use default values
+            self.a, self.b, self.mode = 0, 1, 0.5
+        else:
+            self.a = min(self.data)
+            self.b = max(self.data)
+            self.calculate_mode()
+
+        return self.a, self.b, self.mode
+
+    def calculate_mode(self, round_to=2):
+        """
+        Calculates the mode of a dataset
+        If no single mode, it will approximate the mode using the mean
+
+        Args:
+            round_to (int): Round the mode value. [Default value: 2]
+
+        Returns:
+            float: mode of data
+        """
+        data = Counter(self.data)
+        dict_data = dict(data)
+
+        # Create list of modes from data
+        mode = [k for k, v in dict_data.items() if v == max(list(data.values()))]
+
+        if len(mode) == 1:
+            self.mode = mode[0]
+        else:
+            # Multiple modes
+            # Approximates mode using mean
+            self.mode = sum(self.data) / len(self.data)
```

Comment:
Do you have a reference for this? Even if multiple modes exist, isn't the mode still the same? I might be wrong but I'd appreciate if you could link a source confirming that.

Reply:
From my understanding, mode is the numbers with the highest frequency appearing within data, and could have 0, 1, 1+ modes. For example, with [1, 2, 2, 3, 3, 4], the mode would be both 2 and 3. For triangular distribution, I believe you can only have 1 mode, otherwise it's not a triangle, so rather than raising an error, I use an approximated mode.
If you have other suggestions for what to do in multi-modal cases, let me know

Reply:
I think it's a better idea to let the user know that there are multiple modes which is a violation for triangular distribution rather than returning an arbitrary number(in this case the mean) as the mode. This is because the user needs to know there's a problem, returning the average doesn't indicate that. 

We could also send the list of modes in the error message.

Reply:
Sure, after the data file is read and I calculate the minimum, maximum, and mode, during that mode calculation, if it returns an array of elements I can send an error message. Do you want me to create a custom Exception for it as well? That way it can throw the Exception when it occurs which will be relayed to the user and prevent them from continuing with the Triangular distribution

Reply:
Yes, that would be good. Please go ahead and make the required changes.